# Segmenting and Clustering Neighborhoods in Toronto
In the following program there is a geographic analysis of Toronto city. It is divided in three parts:
1. Web scraping: obtain the database of the neighborhoods needed for the analysis
2. Coordinates: obtain the coordinates of each neighborhood and venue. Create a dataframe with the characteristics of each neighborhood according to the frequency of the different categories of venues. 
3. K-means to cluster neighborhoods: use the k-means algorithm to cluster the neighborhoods into 5 groups as of their characterization in the previous step. Display the results.

## 1. Web Scraping

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

From the wikipedia page extract all the dataframes:

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
df = pd.read_html(url, header=0)
len(df)

3

Select the dataframe of interest and display the first 5 values:

In [3]:
print(len(df[0]["Neighborhood"].unique()))
df_t1 = df[0]
#type(df_t1)
print(df_t1.shape)
df_t1.head()

100
(180, 3)


Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Delete the rows with "Not assigned" for the boroughs. Combine the neighborhood names according to its postal code. One thing to highlight is that there are less unique neighborhood values (i.e. less neighborhood names) than unique postal code values. This means that: i) one neighborhood has two postal codes; or ii) there are different neighborhoods with the same name. _**For this reason, great part of the data analysis is done with the postal code identifier instead of the neighborhood name**_:

In [48]:
df_t2 = df_t1[df_t1["Borough"] != "Not assigned"]
print("The number of unique postal codes:",len(df_t2["Postal Code"].unique())) 
print("The number of unique neighborhoods:",len(df_t2["Neighborhood"].unique()))
df_t2.sort_values("Postal Code", ascending=True)
df_t2.reset_index(inplace=True, drop=True)
df_t2.head()

The number of unique postal codes: 103
The number of unique neighborhoods: 99


Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Here I verify if there is any "Not assigned" value left in Borough or Neighborhood:

In [49]:
if "Not assigned" in df_t2["Borough"].unique():
    print("clean")
else: print("ok")

if "Not assigned" in df_t2["Neighborhood"].unique():
    print("clean")
else: print("ok")

ok
ok


In [50]:
print("The number of rows of the dataset is:",df_t2.shape[0],"and of columns:",df_t2.shape[1])

The number of rows of the dataset is: 103 and of columns: 3


## 2. Coordinates
In this section I add the coordinates information to each neighborhood, according to its postal code.

In this first try is the code necessary to obtain the latitude and longitude from the `geocoder` library.

In [7]:
#! pip install geocoder
#import geocoder
#column_names = ["Postal Code","Borough","Neighborhood","Latitude","Longitude"]
#nh = pd.DataFrame(columns=column_names)
#lat_lng_coords = None
#i = 0
#h= 0
#for post, bor, nei in zip(df_t2["Postal Code"],df_t2["Borough"],df_t2["Neighborhood"]):
#    lat_lng_coords = None
#    h=0
#    while(lat_lng_coords is None):
#        g = geocoder.google('{}, Toronto, Ontario'.format(post))
#        lat_lng_coords = g.latlng
#        h+=1
#        print(h)
#        if h==50: 
#            lat_lng_coords = ["None","None"]
            
#    latitude = lat_lng_coords[0]
#     longitude = lat_lng_coords[1]
#    nh = nh.append({'Postal Code': post,
#                   'Borough': bor,
#                   'Neighborhood': nei,
#                   'Latitude':latitude,
#                   'Longitude':longitude}, ignore_index = True)
#    print(i)
#    i+=1

After many iterations, this api wasn't able to produce the information needed, so I used an external database:

In [8]:
df_lat_lng = pd.read_csv("Geospatial_Coordinates.csv")
print(df_lat_lng.shape)
df_lat_lng.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Here I merge dataframes by "Postal Code":

In [9]:
df_t3 = pd.merge(left=df_t2, right=df_lat_lng, how="left", left_on="Postal Code", right_on = "Postal Code")
print(df_t3.shape)
#print(df_t3["Neighborhood"].unique())
df_t3.head()

(103, 5)


Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Here is use the `geocode` library to get the coordinates of Toronto City:

In [10]:
address = "Toronto, ON"

geolocator = Nominatim(user_agent="t_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

print("The geographical coordinate of Toronto asre {}, {}.".format(latitude,longitude))

The geographical coordinate of Toronto asre 43.6534817, -79.3839347.


The folowing script yields the map of Toronto with the location of each neighborhood (signaled by the markers):

In [13]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighborhood in zip(df_t3['Latitude'], df_t3['Longitude'], df_t3['Borough'], df_t3['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='midnightblue',
        fill=True,
        fill_color='maroon',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [14]:
CLIENT_ID = 'SECRET' # your Foursquare ID
CLIENT_SECRET = 'SECRET' # your Foursquare Secret
VERSION = '20200606' # Foursquare API version

Generate the function that makes a list with the name and coordinates of the nearby locations of each postal code (with limit of 200 locations for each neighborhood and a radius of 1km):

In [33]:
LIMIT = 200
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        try:
            results = requests.get(url).json()["response"]['groups'][0]['items']
        except: continue
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [34]:
toronto_venues = getNearbyVenues(names=df_t3["Postal Code"],
                                latitudes= df_t3["Latitude"],
                                longitudes=df_t3["Longitude"])
toronto_venues.shape

(4881, 7)

In the following dataframe is presented the first 5 venues found with the **foursquare API**:

In [35]:
print(toronto_venues.shape)
toronto_venues.head()


(4881, 7)


Postal Code  Neighborhood Latitude  Neighborhood Longitude  \
0         M3A              43.753259              -79.329656   
1         M3A              43.753259              -79.329656   
2         M3A              43.753259              -79.329656   
3         M3A              43.753259              -79.329656   
4         M3A              43.753259              -79.329656   

                      Venue  Venue Latitude  Venue Longitude  \
0           Allwyn's Bakery       43.759840       -79.324719   
1           Brookbanks Park       43.751976       -79.332140   
2               Tim Hortons       43.760668       -79.326368   
3         Bruno's valu-mart       43.746143       -79.324630   
4  High Street Fish & Chips       43.745260       -79.324949   

         Venue Category  
0  Caribbean Restaurant  
1                  Park  
2                  Café  
3         Grocery Store  
4     Fish & Chips Shop

In [36]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 331 uniques categories.


The following script generates a table that identifies each venue according to its category (there are 331 different categories):

In [37]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
cols = list(toronto_onehot.columns)

# add neighborhood column back to dataframe
toronto_onehot['Postal Code'] = toronto_venues['Postal Code'] 

fixed_columns = [toronto_onehot.columns[-1]]+ list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

In [39]:
print(toronto_onehot.shape)
print(len(toronto_onehot["Postal Code"].unique()))
toronto_onehot.head()

(4881, 332)
101


Postal Code  Accessories Store  Afghan Restaurant  African Restaurant  \
0         M3A                  0                  0                   0   
1         M3A                  0                  0                   0   
2         M3A                  0                  0                   0   
3         M3A                  0                  0                   0   
4         M3A                  0                  0                   0   

   Airport  American Restaurant  Amphitheater  Animal Shelter  Antique Shop  \
0        0                    0             0               0             0   
1        0                    0             0               0             0   
2        0                    0             0               0             0   
3        0                    0             0               0             0   
4        0                    0             0               0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Dealership  Auto Garage  Auto Workshop  \
0                   0                0            0              0   
1                   0                0            0              0   
2                   0                0            0              0   
3                   0                0            0              0   
4                   0                0            0              0   

   Automotive Shop  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  \
0                0          0           0                0           0   
1                0          0           0                0           0   
2                0          0           0                0           0   
3                0          0           0                0           0   
4                0          0           0                0           0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  \
0       0     0    0               0                 0                   0   
1       0     0    0               0                 0                   0   
2       0     0    0               0                 0                   0   
3       0     0    0               0                 0                   0   
4       0     0    0               0                 0                   0   

   Beach  Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0      0          0         0           0                   0          0   
1      0          0         0           0                   0          0   
2      0          0         0           0                   0          0   
3      0          0         0           0                   0          0   
4      0          0         0           0                   0          0   

   Bistro  Bookstore  Botanical Garden  Boutique  Bowling Alley  \
0       0          0                 0         0              0   
1       0          0                 0         0              0   
2       0          0                 0         0              0   
3       0          0                 0         0              0   
4       0          0                 0         0              0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  Bridge  \
0                     0               0        0            0       0   
1                     0               0        0            0       0   
2                     0               0        0            0       0   
3                     0               0        0            0       0   
4                     0               0  

This dataframe contains the mean of the frequency of occurrence of each category:

In [40]:
toronto_g = toronto_onehot.groupby("Postal Code").mean().reset_index()
print(toronto_g.shape)
toronto_g.head()

(101, 332)


Postal Code  Accessories Store  Afghan Restaurant  African Restaurant  \
0         M1C                0.0                0.0                 0.0   
1         M1E                0.0                0.0                 0.0   
2         M1G                0.0                0.0                 0.0   
3         M1H                0.0                0.0                 0.0   
4         M1J                0.0                0.0                 0.0   

   Airport  American Restaurant  Amphitheater  Animal Shelter  Antique Shop  \
0      0.0                  0.0           0.0             0.0           0.0   
1      0.0                  0.0           0.0             0.0           0.0   
2      0.0                  0.0           0.0             0.0           0.0   
3      0.0                  0.0           0.0             0.0           0.0   
4      0.0                  0.0           0.0             0.0           0.0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0       0.0          0.0         0.0                  0.0          0.000000   
1       0.0          0.0         0.0                  0.0          0.000000   
2       0.0          0.0         0.0                  0.0          0.000000   
3       0.0          0.0         0.0                  0.0          0.032258   
4       0.0          0.0         0.0                  0.0          0.000000   

   Athletics & Sports  Auto Dealership  Auto Garage  Auto Workshop  \
0            0.000000              0.0          0.0            0.0   
1            0.000000              0.0          0.0            0.0   
2            0.000000              0.0          0.0            0.0   
3            0.032258              0.0          0.0            0.0   
4            0.000000              0.0          0.0            0.0   

   Automotive Shop  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  \
0              0.0        0.0         0.0              0.0         0.0   
1              0.0        0.0         0.0              0.0         0.0   
2              0.0        0.0         0.0              0.0         0.0   
3              0.0        0.0         0.0              0.0         0.0   
4              0.0        0.0         0.0              0.0         0.0   

     Bakery      Bank  Bar  Baseball Field  Baseball Stadium  \
0  0.000000  0.000000  0.0             0.0               0.0   
1  0.000000  0.080000  0.0             0.0               0.0   
2  0.000000  0.000000  0.0             0.0               0.0   
3  0.096774  0.064516  0.0             0.0               0.0   
4  0.000000  0.000000  0.0             0.0               0.0   

   Basketball Stadium  Beach  Beach Bar  Beer Bar  Beer Store  \
0                 0.0    0.0        0.0       0.0        0.00   
1                 0.0    0.0        0.0       0.0        0.04   
2                 0.0    0.0        0.0       0.0        0.00   
3                 0.0    0.0        0.0       0.0        0.00   
4                 0.0    0.0        0.0       0.0        0.00   

   Belgian Restaurant  Bike Shop  Bistro  Bookstore  Botanical Garden  \
0                 0.0        0.0     0.0        0.0               0.0   
1                 0.0        0.0     0.0        0.0               0.0   
2                 0.0        0.0     0.0        0.0               0.0   
3                 0.0        0.0     0.0        0.0               0.0   
4                 0.0        0.0     0.0        0.0               0.0   

   Boutique  Bowling Alley  Brazilian Restaurant  Breakfast Spot  Brewery  \
0       0.0       0.000000                   0.0             0.2      0.0   
1       0.0       0.000000                   0.0             0.0      0.0   
2       0.0       0.000000                   0.0             0.0      0.0   
3       0.0       0.000000                   0.0             0.0      0.0   
4       0.0       0.076923                   0.0             0.0      0.0   

   Bridal Shop  Bridge  Bubble Tea Shop  Buffet  Burger J

Here are the top 5 most frequent venue categories for each postal code location:

In [41]:
num_top_venues = 5

for hood in toronto_g['Postal Code']:
    print("----"+hood+"----")
    temp = toronto_g[toronto_g['Postal Code'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M1C----
                venue  freq
0                Park   0.2
1      Breakfast Spot   0.2
2        Burger Joint   0.2
3  Italian Restaurant   0.2
4          Playground   0.2


----M1E----
                  venue  freq
0           Pizza Place  0.12
1  Fast Food Restaurant  0.08
2                  Bank  0.08
3            Restaurant  0.08
4           Coffee Shop  0.08


----M1G----
                  venue  freq
0                  Park  0.22
1           Coffee Shop  0.22
2  Fast Food Restaurant  0.11
3              Pharmacy  0.11
4    Chinese Restaurant  0.11


----M1H----
               venue  freq
0             Bakery  0.10
1        Coffee Shop  0.10
2               Bank  0.06
3  Indian Restaurant  0.06
4        Gas Station  0.06


----M1J----
                  venue  freq
0        Ice Cream Shop  0.15
1         Bowling Alley  0.08
2        Sandwich Place  0.08
3  Fast Food Restaurant  0.08
4           Coffee Shop  0.08


----M1K----
                venue  freq
0  Chinese Restauran

In [42]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

This dataframe shows the top 20 mosto common venue category for each neighborhood:

In [43]:
num_top_venues = 20

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postal Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Postal Code'] = toronto_g['Postal Code']

for ind in np.arange(toronto_g.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_g.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0         M1C    Italian Restaurant        Breakfast Spot   
1         M1E           Pizza Place           Coffee Shop   
2         M1G                  Park           Coffee Shop   
3         M1H                Bakery           Coffee Shop   
4         M1J        Ice Cream Shop           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Playground          Burger Joint                  Park   
1  Fast Food Restaurant            Restaurant                  Bank   
2    Chinese Restaurant     Indian Restaurant  Fast Food Restaurant   
3                  Bank     Indian Restaurant           Gas Station   
4         Train Station            Restaurant  Fast Food Restaurant   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0                   Zoo  Eastern European Restaurant     Electronics Store   
1   Filipino Restaurant                     Pharmacy         Smoothie Shop   
2              Pharmacy            Mobile Phone Shop          Dance Studio   
3         Grocery Store         Gym / Fitness Center          Burger Joint   
4   Japanese Restaurant                Bowling Alley     Convenience Store   

  9th Most Common Venue       10th Most Common Venue 11th Most Common Venue  \
0  Ethiopian Restaurant                  Event Space     Falafel Restaurant   
1     Food & Drink Shop                   Sports Bar             Beer Store   
2        Farmers Market  Eastern European Restaurant      Electronics Store   
3  Fast Food Restaurant                  Pizza Place    Sporting Goods Shop   
4        Sandwich Place                  Pizza Place          Women's Store   

  12th Most Common Venue 13th Most Common Venue       14th Most Common Venue  \
0                   Farm         Farmers Market                        Field   
1           Liquor Store         Sandwich Place             Greek Restaurant   
2   Ethiopian Restaurant            Event Space           Falafel Restaurant   
3    Fried Chicken Joint   Caribbean Restaurant                  Music Store   
4          Grocery Store    Dumpling Restaurant  Eastern European Restaurant   

  15th Most Common Venue 16th Most Common Venue 17th Most Common Venue  \
0   Fast Food Restaurant              Drugstore    Filipino Restaurant   
1          Moving Target     Chinese Restaurant    Fried Chicken Joint   
2                   Farm               Creperie              Drugstore   
3     Chinese Restaurant        Thai Restaurant               Bus Line   
4      Electronics Store              Drugstore   Ethiopian Restaurant   

  18th Most Common Venue 19th Most Common Venue 20th Most Common Venue  
0        Fireworks Store      Fish & Chips Shop            Fish Market  
1         Discount Store            Supermarket   Ethiopian Restaurant  
2                  Field    Filipino Restaurant        Fireworks Store  
3            Wings Joint     Athletics & Sports               Pharmacy  
4            Event Space                    Zoo     Falafel Restaurant

## 3. K-means to cluster Neighborhoods
The following scripts runs a *k-means* algorithm to cluster in five groups all the neighborhoods according to the frequency of the different categories of venues in each neighborhood:

In [44]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_g.drop('Postal Code', axis = 1)
toronto_grouped_clustering.head()
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
check1 = list(kmeans.labels_)
len(check1)

101

In [45]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_t3

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Postal Code'), on='Postal Code')
print(toronto_merged["Cluster Labels"].unique())
toronto_merged.head() # check the last columns!
toronto_merged.replace("nan", np.nan, inplace = True)
toronto_merged.dropna(inplace=True)
print(toronto_merged["Cluster Labels"].unique())

[ 2.  3.  0. nan  1.  4.]
[2. 3. 0. 1. 4.]


### Clustered Neighborhoods Map
This map shows the 5 different clusters of neighborhoods, indicating which neighborhoods are more similar between them:

In [47]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
#print(colors_array)
rainbow = [colors.rgb2hex(i) for i in colors_array]
#print(rainbow)
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    #print(cluster)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)],
        fill=True,
        fill_color=rainbow[int(cluster)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### __*Cluster 1*__

In [51]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5      Etobicoke             0.0              Pharmacy     Convenience Store   
12   Scarborough             0.0    Italian Restaurant        Breakfast Spot   
50    North York             0.0    Italian Restaurant                  Bank   
57    North York             0.0           Golf Course      Storage Facility   
58   Scarborough             0.0                  Park       Thai Restaurant   
66    North York             0.0                  Park            Restaurant   
88     Etobicoke             0.0                  Park            Restaurant   
101    Etobicoke             0.0    Italian Restaurant                  Park   

    3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
5                    Bank                       Bakery            Playground   
12             Playground                 Burger Joint                  Park   
50                 Bakery               Medical Center           Pizza Place   
57            Gas Station                         Park        Discount Store   
58                   Farm                        Diner          Skating Rink   
66            Coffee Shop                          Gym         Grocery Store   
88      Indian Restaurant                 Skating Rink    Italian Restaurant   
101         Shopping Mall  Eastern European Restaurant        Ice Cream Shop   

    6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
5             Golf Course                Shopping Mall                  Park   
12                    Zoo  Eastern European Restaurant     Electronics Store   
50          Shopping Mall            Electronics Store              Pharmacy   
57           Intersection                       Bakery     Convenience Store   
58                    Gym                         Café              Gym Pool   
66      French Restaurant                  Gas Station          Intersection   
88    Fried Chicken Joint                Grocery Store                Bakery   
101              Bus Stop                          Zoo   Dumpling Restaurant   

    9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
5                    Café          Grocery Store      Electronics Store   
12   Ethiopian Restaurant            Event Space     Falafel Restaurant   
50    Arts & Crafts Store                   Park                    Zoo   
57    Filipino Restaurant            Fish Market    Dumpling Restaurant   
58             Restaurant  General Entertainment         Ice Cream Shop   
66       Dentist's Office            Golf Course       Business Service   
88                   Café               Pharmacy   Fast Food Restaurant   
101     Electronics Store   Ethiopian Restaurant            Event Space   

          12th Most Common Venue       13th Most Common Venue  \
5                      Drugstore          Dumpling Restaurant   
12                          Farm               Farmers Market   
50           Dumpling Restaurant  Eastern European Restaurant   
57   Eastern European Restaurant            Electronics Store   
58               College Stadium            Convenience Store   
66                  Tennis Court            Convenience Store   
88                           Pub                 Liquor Store   
101           Falafel Restaurant               Farmers Market   

          14th Most Common Venue 15th Most Common Venue  \
5    Eastern European Restaurant            Event Space   
12                         Field   Fast Food Restaurant   
50          Ethiopian Restaurant            Event Space   
57          Ethiopian Restaurant            Flea Market   
58                Farmers Market        Fireworks Store   
66                    Playground                Dog Run   
88           American Restaurant            Pizza Place   
101                         Farm             Donut Shop   

    16th Most Common Venue 17th Most Common Venue 18

### __*Cluster 2*__

In [52]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             1.0                  Park                  Pool   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45                   Zoo                  Farm             Drugstore   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
45   Dumpling Restaurant  Eastern European Restaurant     Electronics Store   

   9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
45  Ethiopian Restaurant            Event Space     Falafel Restaurant   

   12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
45         Farmers Market       Doner Restaurant   Fast Food Restaurant   

   15th Most Common Venue 16th Most Common Venue 17th Most Common Venue  \
45                  Field    Filipino Restaurant        Fireworks Store   

   18th Most Common Venue 19th Most Common Venue 20th Most Common Venue  
45      Fish & Chips Shop            Fish Market            Flea Market

### _**Cluster 3**_

In [53]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
0    North York             2.0                    Park   
8     East York             2.0                 Brewery   
10   North York             2.0           Grocery Store   
11    Etobicoke             2.0                    Park   
16         York             2.0             Pizza Place   
18  Scarborough             2.0             Pizza Place   
21         York             2.0                Pharmacy   
22  Scarborough             2.0                    Park   
26  Scarborough             2.0                  Bakery   
27   North York             2.0                Pharmacy   
28   North York             2.0                    Bank   
32  Scarborough             2.0          Ice Cream Shop   
38  Scarborough             2.0      Chinese Restaurant   
39   North York             2.0                    Bank   
44  Scarborough             2.0            Intersection   
46   North York             2.0                    Park   
49   North York             2.0             Coffee Shop   
51  Scarborough             2.0                   Beach   
56         York             2.0  Furniture / Home Store   
60   North York             2.0                   Hotel   
63         York             2.0             Coffee Shop   
70    Etobicoke             2.0             Pizza Place   
71  Scarborough             2.0           Grocery Store   
72   North York             2.0                Pharmacy   
77    Etobicoke             2.0                Pharmacy   
78  Scarborough             2.0      Chinese Restaurant   
82  Scarborough             2.0             Coffee Shop   
85  Scarborough             2.0      Chinese Restaurant   
89    Etobicoke             2.0             Pizza Place   
90  Scarborough             2.0      Chinese Restaurant   
93    Etobicoke             2.0          Discount Store   

          2nd Most Common Venue      3rd Most Common Venue  \
0             Convenience Store                   Bus Stop   
8                   Pizza Place       Gym / Fitness Center   
10         Fast Food Restaurant                Pizza Place   
11                  Pizza Place         Mexican Restaurant   
16                  Coffee Shop          Convenience Store   
18                  Coffee Shop       Fast Food Restaurant   
21                         Park                   Bus Stop   
22                  Coffee Shop         Chinese Restaurant   
26                  Coffee Shop                       Bank   
27                  Coffee Shop                       Park   
28                         Park                Pizza Place   
32                  Coffee Shop              Train Station   
38                  Coffee Shop                Pizza Place   
39                  Gas Station              Grocery Store   
44                     Bus Line                     Bakery   
46                  Coffee Shop              Shopping Mall   
49           Dim Sum Restaurant          Convenience Store   
51                  Pizza Place             Ice Cream Shop   
56                Grocery Store               Dessert Shop   
60                  Gas Station                Coffee Shop   
63                  Pizza Place                    Brewery   
70                  Gas Station             Sandwich Place   
71                  Pizza Place  Middle Eastern Restaurant   
72  Eastern European Restaurant                       Park   
77                  Pizza Place               Intersection   
78                Shopping Mall             Sandwich Place   
82                         Bank               Intersection   
85                  Pizza Place          Korean Restaurant   
89                Grocery Store             Sandwich Place   
90                  Coffee Shop                       Bank   
93                  Pizza Place          Convenience Store   

     4th Most Common Venue   5th Most Common Venue 6th Most Common Venue  \
0                 Pharmacy           Shopping Mall     Foo

### _**Cluster 4**_

In [54]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
1          North York             3.0             Coffee Shop   
2    Downtown Toronto             3.0             Coffee Shop   
3          North York             3.0    Fast Food Restaurant   
4    Downtown Toronto             3.0             Coffee Shop   
7          North York             3.0             Coffee Shop   
9    Downtown Toronto             3.0             Coffee Shop   
13         North York             3.0              Restaurant   
14          East York             3.0             Coffee Shop   
15   Downtown Toronto             3.0             Coffee Shop   
17          Etobicoke             3.0             Coffee Shop   
19       East Toronto             3.0                     Pub   
20   Downtown Toronto             3.0             Coffee Shop   
23          East York             3.0             Coffee Shop   
24   Downtown Toronto             3.0             Coffee Shop   
25   Downtown Toronto             3.0       Korean Restaurant   
29          East York             3.0             Coffee Shop   
30   Downtown Toronto             3.0             Coffee Shop   
31       West Toronto             3.0                    Café   
33         North York             3.0             Coffee Shop   
34         North York             3.0             Coffee Shop   
35          East York             3.0             Coffee Shop   
36   Downtown Toronto             3.0             Coffee Shop   
37       West Toronto             3.0                    Café   
40         North York             3.0             Coffee Shop   
41       East Toronto             3.0        Greek Restaurant   
42   Downtown Toronto             3.0             Coffee Shop   
43       West Toronto             3.0                    Café   
47       East Toronto             3.0       Indian Restaurant   
48   Downtown Toronto             3.0             Coffee Shop   
52         North York             3.0       Korean Restaurant   
54       East Toronto             3.0             Coffee Shop   
55         North York             3.0      Italian Restaurant   
59         North York             3.0             Coffee Shop   
61    Central Toronto             3.0             College Gym   
62    Central Toronto             3.0      Italian Restaurant   
64               York             3.0           Train Station   
65        Scarborough             3.0  Furniture / Home Store   
67    Central Toronto             3.0             Coffee Shop   
68    Central Toronto             3.0             Coffee Shop   
69       West Toronto             3.0                    Café   
73    Central Toronto             3.0             Coffee Shop   
74    Central Toronto             3.0      Italian Restaurant   
75       West Toronto             3.0                    Café   
76        Mississauga             3.0             Coffee Shop   
79    Central Toronto             3.0             Coffee Shop   
80   Downtown Toronto             3.0                    Café   
81       West Toronto             3.0             Coffee Shop   
83    Central Toronto             3.0             Coffee Shop   
84   Downtown Toronto             3.0                    Café   
86    Central Toronto             3.0             Coffee Shop   
87   Downtown Toronto             3.0             Coffee Shop   
91   Downtown Toronto             3.0             Coffee Shop   
92   Downtown Toronto             3.0             Coffee Shop   
94          Etobicoke             3.0                   Hotel   
96   Downtown Toronto             3.0                    Café   
97   Downtown Toronto             3.0             Coffee Shop   
98          Etobicoke             3.0             Coffee Shop   
99   Downtown Toronto             3.0             Coffee Shop   
100      East Toronto             3.0                    Park   
102         Etobicoke             3.0              Restaurant   

      2nd Most Common Venue          3rd Most Co

### _**Cluster 5**__

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue 2nd Most Common Venue  \
53  North York             4.0  Vietnamese Restaurant            Food Truck   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
53        Baseball Field        Farmers Market   Dumpling Restaurant   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
53  Eastern European Restaurant     Electronics Store  Ethiopian Restaurant   

   9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
53           Event Space     Falafel Restaurant                   Farm   

   12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
53   Fast Food Restaurant             Donut Shop                  Field   

   15th Most Common Venue 16th Most Common Venue 17th Most Common Venue  \
53    Filipino Restaurant        Fireworks Store      Fish & Chips Shop   

   18th Most Common Venue 19th Most Common Venue 20th Most Common Venue  
53            Fish Market            Flea Market            Flower Shop